### PR TITLE
Add tracking of player mods, score, passed status

### PR DESCRIPTION
**Pull request by matte-ek**
This PR adds mod, score and passed tracking for every player. It will also handle mods set within the room, such as Freemod or if the room is forcing any other mods.

Depends on https://github.com/hburn7/BanchoSharp/pull/6

### CODE REVIEW

File: BanchoSharp/Multiplayer/MultiplayerMods.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    SpunOut = 1 << 14
+}
+
+public static class ModsExtensions
+{
+
+    /// <summary>
+    /// Returns a string with a shorter format, example: "HDDT"
+    /// </summary>
+    public static string ToShortString(this Mods mods, bool showFreemod = true)
```
Comment (hburn7) [inline BanchoSharp/Multiplayer/MultiplayerMods.cs]: I think the name `ToAbbreviatedForm` works better here.

File: BanchoSharp/Multiplayer/MultiplayerMods.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        var ret = "";
```
Comment (hburn7) [inline BanchoSharp/Multiplayer/MultiplayerMods.cs]: I would rather a `StringBuilder` object be used here, string appending takes up resources.

File: BanchoSharp/Multiplayer/MultiplayerLobby.cs [hunk 3]
```diff
@@ -408,6 +373,195 @@ private void UpdateLobbyFromBanchoBotSettingsResponse(string banchoBotResponse)
 
 			OnPlayerSlotMove?.Invoke(new PlayerSlotMoveEventArgs(player, previousSlot, slotNum));
 		}
+		else if (banchoBotResponse.StartsWith("Players: "))
```
Comment (hburn7) [inline BanchoSharp/Multiplayer/MultiplayerLobby.cs]: See #6

File: BanchoSharp/Multiplayer/MultiplayerLobby.cs [hunk 4]
```diff
@@ -408,6 +373,195 @@ private void UpdateLobbyFromBanchoBotSettingsResponse(string banchoBotResponse)
...
+		else if (banchoBotResponse.StartsWith("Slot "))
+		{
+			// Find the digit(s) within the first 8 characters
+			var slot = int.Parse((banchoBotResponse[..8].Where(c => char.IsDigit(c)).ToArray()));
+
+			// Find the first ' ' after the URL, since the URL is not padded with any spaces.
+			var playerNameBegin = banchoBotResponse.IndexOf(' ', banchoBotResponse.IndexOf("/u/")) + 1;
+
+			var name = banchoBotResponse.Substring(playerNameBegin, 16).TrimEnd();
+			var info = banchoBotResponse.Length > (playerNameBegin + 16) ? banchoBotResponse[(playerNameBegin + 16)..] : null;
```
Comment (hburn7) [inline BanchoSharp/Multiplayer/MultiplayerLobby.cs:401]: More descriptive name needed, not sure what this is.

File: BanchoSharp/Multiplayer/MultiplayerLobby.cs [hunk 5]
```diff
@@ -408,6 +373,195 @@ private void UpdateLobbyFromBanchoBotSettingsResponse(string banchoBotResponse)
...
+		}
+		else if (banchoBotResponse.StartsWith("Slot "))
+		{
+			// Find the digit(s) within the first 8 characters
+			var slot = int.Parse((banchoBotResponse[..8].Where(c => char.IsDigit(c)).ToArray()));
+
+			// Find the first ' ' after the URL, since the URL is not padded with any spaces.
+			var playerNameBegin = banchoBotResponse.IndexOf(' ', banchoBotResponse.IndexOf("/u/")) + 1;
+
+			var name = banchoBotResponse.Substring(playerNameBegin, 16).TrimEnd();
```
Comment (hburn7) [inline BanchoSharp/Multiplayer/MultiplayerLobby.cs]: Please specify what object name is referring to in the variable name. e.g. `playerName`

File: BanchoSharp/Multiplayer/MultiplayerPlayer.cs [hunk 6]
```diff
@@ -7,19 +7,24 @@ public enum TeamColor
...
 {
 	public string Name { get; }
 	public TeamColor Team { get; set; }
 	public int Slot { get; set; }
+	
+	// The mods the player is using, these only get updated after "!mp settings" has been ran.
+	public Mods Mods { get; set; }
+
+	public int? Score { get; set; } = null;
+	public bool? Passed { get; set; } = null;
```
Comment (hburn7) [inline BanchoSharp/Multiplayer/MultiplayerPlayer.cs]: The default value of a nullable value type is null, so no need to declare that here.

**Review by hburn7 — CHANGES_REQUESTED** [review]
After these changes are made, every single check in that spaghetti monster of an if statement (my fault for starting it!) will need to be extracted to it's own boolean method. Currently this isn't testable.

edit: resolved in d1191de9319f9d1e464c644d5a2602300241376b

**Comment by hburn7** [comment]
I knew there'd be conflicts, please merge v1.2.0 into this PR branch.

**Review by matte-ek — COMMENTED** [review]
Since only auto pilot seems to have a 'special' name, I think an edge-case check is fine. If more mods comes up with different names, we should probably consider a better solution.

**Comment by hburn7** [comment]
This PR is stale, will close and have asked @matte-ek to reopen.

**Pull request closed without merging**